src/software/simulation/simulator.cpp

```diff
@@ -11,63 +11,88 @@ extern "C"
+{
+    physics_world.addYellowRobots(robots);
+    for (auto physics_robot : physics_world.getYellowPhysicsRobots())
     {
-        RobotStateWithId state{.id          = robot.id(),
-                               .robot_state = robot.currentState().robotState()};
-        blue_robots.emplace_back(state);
+        auto simulator_robot = std::make_shared<SimulatorRobot>(physics_robot);
+        auto firmware_robot  = SimulatorRobotSingleton::createFirmwareRobot();
+        auto firmware_ball   = SimulatorBallSingleton::createFirmwareBall();
+        // Release ownership of the pointers so the firmware_world can take ownership
+        FirmwareWorld_t* firmware_world_raw =
+            app_firmware_world_create(firmware_robot.release(), firmware_ball.release());
+        auto firmware_world =
+            std::shared_ptr<FirmwareWorld_t>(firmware_world_raw, FirmwareWorldDeleter());
+
+        yellow_simulator_robots.insert(std::make_pair(simulator_robot, firmware_world));
     }
-    physics_world.addBlueRobots(blue_robots);
+}
 
-    for (auto physics_robot : physics_world.getFriendlyPhysicsRobots())
+void Simulator::addBlueRobots(const std::vector<RobotStateWithId>& robots)
+{
+    physics_world.addBlueRobots(robots);
+    for (auto physics_robot : physics_world.getBluePhysicsRobots())
     {
         auto simulator_robot = std::make_shared<SimulatorRobot>(physics_robot);
-
-        auto firmware_robot = SimulatorRobotSingleton::createFirmwareRobot();
-        auto firmware_ball  = SimulatorBallSingleton::createFirmwareBall();
+        auto firmware_robot  = SimulatorRobotSingleton::createFirmwareRobot();
+        auto firmware_ball   = SimulatorBallSingleton::createFirmwareBall();
         // Release ownership of the pointers so the firmware_world can take ownership
         FirmwareWorld_t* firmware_world_raw =
             app_firmware_world_create(firmware_robot.release(), firmware_ball.release());
         auto firmware_world =
             std::shared_ptr<FirmwareWorld_t>(firmware_world_raw, FirmwareWorldDeleter());
 
-        simulator_robots.insert(std::make_pair(simulator_robot, firmware_world));
+        blue_simulator_robots.insert(std::make_pair(simulator_robot, firmware_world));
     }
-
-    simulator_ball = std::make_shared<SimulatorBall>(physics_world.getPhysicsBall());
 }
 
-void Simulator::stepSimulation(const Duration& time_step)
+void Simulator::setYellowRobotPrimitives(ConstPrimitiveVectorPtr primitives)
 {
-    SimulatorBallSingleton::setSimulatorBall(simulator_ball);
-    for (auto& iter : simulator_robots)
+    if (!primitives)
     {
-        auto simulator_robot = iter.first;
-        auto firmware_world  = iter.second;
-        SimulatorRobotSingleton::setSimulatorRobot(simulator_robot);
-        SimulatorRobotSingleton::runPrimitiveOnCurrentSimulatorRobot(firmware_world);
+        return;
     }
 
-    physics_world.stepSimulation(time_step);
+    SimulatorBallSingleton::setSimulatorBall(simulator_ball);
```

Comment:
Similar comment to @akhilveeraghanta below, what's this for?

Reply:
pretty much same reply as above. Need to do this so all robots are "seeing" the most recent state of the ball in their `firmware_world`. Added a comment.